retrieval/discovery/consul/consul.go

```diff
@@ -0,0 +1,259 @@
+
+		// Check for new services.
+		for name := range srvs {
+			if !cd.shouldWatch(name) {
+				continue
+			}
+			if _, ok := services[name]; ok {
+				continue // We are already watching the service.
+			}
+
+			srv := &consulService{
+				client: cd.client,
+				name:   name,
+				labels: model.LabelSet{
+					serviceLabel:    model.LabelValue(name),
+					datacenterLabel: model.LabelValue(cd.clientDatacenter),
+				},
+				tagSeparator: cd.tagSeparator,
+			}
+
+			wctx, cancel := context.WithCancel(ctx)
+			go srv.watch(wctx, ch)
+
+			services[name] = cancel
+		}
+
+		// Check for removed services.
+		for name, cancel := range services {
+			if _, ok := srvs[name]; !ok {
+				// Call the watch cancelation function.
+				cancel()
+				delete(services, name)
+
+				// Send clearing target group.
+				select {
+				case <-ctx.Done():
+					return
+				case ch <- []*config.TargetGroup{{Source: name}}:
+				}
+			}
+		}
+	}
+}
+
+// consulService contains data belonging to the same service.
+type consulService struct {
+	name         string
+	labels       model.LabelSet
+	client       *consul.Client
+	tagSeparator string
+}
+
+func (srv *consulService) watch(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	catalog := srv.client.Catalog()
+
+	lastIndex := uint64(0)
+	for {
+		nodes, meta, err := catalog.Service(srv.name, "", &consul.QueryOptions{
+			WaitIndex: lastIndex,
+			WaitTime:  watchTimeout,
+		})
+		// Check the context before potentially falling in a continue-loop.
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// Continue.
+		}
+
+		if err != nil {
+			log.Errorf("Error refreshing service %s: %s", srv.name, err)
+			time.Sleep(retryInterval)
+			continue
+		}
+		// If the index equals the previous one, the watch timed out with no update.
+		if meta.LastIndex == lastIndex {
+			continue
+		}
+		lastIndex = meta.LastIndex
+
+		tgroup := config.TargetGroup{
+			Source:  srv.name,
+			Labels:  srv.labels,
+			Targets: make([]model.LabelSet, 0, len(nodes)),
+		}
+
+		for _, node := range nodes {
+			var (
+				addr = fmt.Sprintf("%s:%d", node.Address, node.ServicePort)
+				// We surround the separated list with the separator as well. This way regular expressions
+				// in relabeling rules don't have to consider tag positions.
+				tags = srv.tagSeparator + strings.Join(node.ServiceTags, srv.tagSeparator) + srv.tagSeparator
+			)
+			tgroup.Targets = append(tgroup.Targets, model.LabelSet{
+				model.AddressLabel:  model.LabelValue(addr),
+				addressLabel:        model.LabelValue(node.Address),
+				nodeLabel:           model.LabelValue(node.Node),
+				tagsLabel:           model.LabelValue(tags),
+				serviceAddressLabel: model.LabelValue(node.ServiceAddress),
+				servicePortLabel:    model.LabelValue(strconv.Itoa(node.ServicePort)),
+				serviceIDLabel:      model.LabelValue(node.ServiceID),
+			})
+		}
+		// Check context twice to ensure we always catch cancelation.
+		select {
+		case <-ctx.Done():
```

Comment:
Doesn't this need a default case or otherwise it'd block forever if not canceled?